src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs

```diff
@@ -2872,40 +2882,84 @@ private int[] SerializeMethodBodies(BlobBuilder ilBuilder, PdbWriter nativePdbWr
             return bodyOffsets;
         }
 
-        private int SerializeMethodBody(MethodBodyStreamEncoder encoder, IMethodBody methodBody, StandaloneSignatureHandle localSignatureHandleOpt, ref UserStringHandle mvidStringHandle, ref Blob mvidStringFixup)
-        {
-            int ilLength = methodBody.IL.Length;
-            var exceptionRegions = methodBody.ExceptionRegions;
-            bool isSmallBody = ilLength < 64 && methodBody.MaxStack <= 8 && localSignatureHandleOpt.IsNil && exceptionRegions.Length == 0;
-
-            // Check if an identical method body has already been serialized. 
+        private int SerializeThrowNullMethodBody(MethodBodyStreamEncoder encoder)
+        {
+            UserStringHandle mvidStringHandle = default(UserStringHandle);
+            Blob mvidStringFixup = default(Blob);
+
+            int offset = SerializeMethodBody(
+                encoder,
+                ThrowNullIL,
+                ThrowNullMaxStack,
+                ImmutableArray<ExceptionHandlerRegion>.Empty,
+                default(StandaloneSignatureHandle),
+                MethodBodyAttributes.None,
+                ref mvidStringHandle,
+                ref mvidStringFixup);
+
+            Debug.Assert(mvidStringHandle.Equals(default(UserStringHandle)) && mvidStringFixup.Equals(default(Blob)));
+            return offset;
+        }
+
+        private int SerializeMethodBody(
+            MethodBodyStreamEncoder encoder,
+            IMethodBody methodBody,
+            StandaloneSignatureHandle localSignatureHandleOpt,
+            ref UserStringHandle mvidStringHandle,
+            ref Blob mvidStringFixup)
+        {
+            return SerializeMethodBody(
+                encoder,
+                methodBody.IL,
+                methodBody.MaxStack,
+                methodBody.ExceptionRegions,
+                localSignatureHandleOpt,
+                (methodBody.LocalsAreZeroed ? MethodBodyAttributes.InitLocals : MethodBodyAttributes.None),
+                ref mvidStringHandle,
+                ref mvidStringFixup);
+        }
```

Comment:
We can avoid all this clutter if we hardcode encoded "throw null" body instead of just the IL.